pkg/volume/csi/csi_test.go

```diff
@@ -0,0 +1,366 @@
+			// unmount volume
+			t.Log("csiTest.VolumeAll Unmouting volume...")
+			volPlug, err = plugMgr.FindPluginBySpec(volSpec)
+			if err != nil || volPlug == nil {
+				t.Fatalf("csiTest.VolumeAll PluginMgr.FindPluginBySpec failed: %v", err)
+			}
+			if volPlug == nil {
+				t.Fatalf("csiTest.VolumeAll volumePlugin is nil")
+			}
+			mounter, err = volPlug.NewMounter(volSpec, pod, volume.VolumeOptions{})
+			if err != nil || mounter == nil {
+				t.Fatalf("csiTest.VolumeAll volPlugin.NewMounter is nil or error: %s", err)
+			}
+
+			unmounter, err := volPlug.NewUnmounter(test.specName, pod.GetUID())
+			if err != nil {
+				t.Fatal("csiTest.VolumeAll volumePlugin.NewUnmounter failed:", err)
+			}
+			csiUnmounter := unmounter.(*csiMountMgr)
+			csiUnmounter.csiClient = csiClient
+
+			if err := csiUnmounter.TearDownAt(mounter.GetPath()); err != nil {
+				t.Fatal("csiTest.VolumeAll unmounter.TearDownAt failed:", err)
+			}
+			t.Log("csiTest.VolumeAll unmounter.TearDownAt done OK for dir:", mounter.GetPath())
+
+			// unmount device
+			t.Log("csiTest.VolumeAll Unmouting device...")
+			devicePlug, err = plugMgr.FindDeviceMountablePluginBySpec(volSpec)
+			if err != nil {
+				t.Fatalf("csiTest.VolumeAll failed to create mountable device plugin: %s", err)
+			}
+
+			var devUnmounter volume.DeviceUnmounter
+			if devicePlug != nil {
+				t.Log("csiTest.VolumeAll found DeviceMountablePlugin, entering device unmouting ...")
+				devMounter, err = devicePlug.NewDeviceMounter()
+				if err != nil {
+					t.Log("csiTest.VolumeAll failed to create new device mounter: ", err)
+				}
+				devUnmounter, err = devicePlug.NewDeviceUnmounter()
+				if err != nil {
+					t.Log("csiTest.VolumeAll failed to create new device unmounter: ", err)
+				}
+
+				if devMounter != nil && devUnmounter != nil {
+					csiDevMounter := devMounter.(*csiAttacher)
+					csiDevUnmounter := devUnmounter.(*csiAttacher)
+					csiDevUnmounter.csiClient = csiClient
+
+					devMountPath, err := csiDevMounter.GetDeviceMountPath(volSpec)
+					if err != nil {
+						t.Fatalf("csiTest.VolumeAll deviceMounter.GetdeviceMountPath failed %s", err)
+					}
+					if err := csiDevUnmounter.UnmountDevice(devMountPath); err != nil {
+						t.Fatalf("csiTest.VolumeAll deviceMounter.UnmountDevice failed: %s", err)
+					}
+					t.Log("csiTest.VolumeAll deviceUmounter.UnmountDevice done OK for path", devMountPath)
+				}
+			} else {
+				t.Log("csiTest.VolumeAll DeviceMountablePluginBySpec did not find a plugin, skipping unmounting.")
+			}
+
+			// detach volume
+			t.Log("csiTest.VolumeAll Detaching volume...")
+			attachPlug, err = plugMgr.FindAttachablePluginBySpec(volSpec)
+			if err != nil {
+				t.Fatalf("csiTest.VolumeAll PluginManager.FindAttachablePluginBySpec failed: %v", err)
+			}
+
+			if attachPlug != nil {
+				volDetacher, err := attachPlug.NewDetacher()
+				if err != nil {
+					t.Log("csiTest.VolumeAll failed to create new detacher: ", err)
```

Comment:
Should this be fatal?

Reply:
Yes.

Reply:
still see this as as a log instead of fatal

Reply:
I missed it, fixed.